Framework: Extract "edit-post" to its own module

youknowriad opened the pull request:
This PR extracts the `edit-post` module to its own module. This step is helpful to be able to experiment with an "edit-template" later.

 - The new module has its own store handling layout state.
 - Extract some more reusable components in `editor` for use in `edit-post` (`BlockSelectionClearer`, `InserterWithShortcuts`...)
 - The "Show block inspector" button proved to be a challenge, because it's layout related. this is fixed by providing a way to extend the Block's menu from outside the `editor` module using a prop.
 - Renamed classNames for the components moved to the new module.

**Testing instructions**

 - This may introduce some unnoticed regressions, heavily testing this PR is necessary.

youknowriad committed:
Framework: Extract edit-post to its own module

youknowriad committed:
Fixing the unit tests

mcsf committed:
Rename 'editor-*' classes to 'edit-post-*' when applicable

youknowriad committed:
Fix e2e tests

youknowriad committed:
Editor with shortcuts: Fix the classnames

youknowriad committed:
Fix Block Inspector Button classname

aduth commented on lib/client-assets.php:
```diff
@@ -159,13 +159,19 @@ function gutenberg_register_scripts_and_styles() {
+	wp_register_script(
+		'wp-edit-post',
+		gutenberg_url( 'edit-post/build/index.js' ),
+		array( 'wp-element', 'wp-components', 'wp-editor', 'wp-i18n', 'wp-date', 'wp-utils' ),
```
`edit-post/index.js` has a reference to jQuery and heartbeat, so this should be made an explicit dependency.

Heartbeat is no longer used as a dependency in `wp-editor`.

aduth commented on webpack.config.js:
```diff
@@ -53,6 +53,7 @@ const entryPointNames = [
 	'i18n',
 	'utils',
 	'data',
+	'edit-post',
```
To be able to export this as `wp.editPost` we should consider mapping packages using [Lodash's `_.camelCase`](https://lodash.com/docs/4.17.4#camelCase).

youknowriad replied:
Might not be possible to configure webpack to do so https://webpack.js.org/configuration/output/#output-library

aduth commented on package.json:
```diff
@@ -104,15 +104,15 @@
 	},
 	"jest": {
 		"collectCoverageFrom": [
-			"(blocks|components|date|editor|element|i18n|data|utils)/**/*.js"
```
We also include a custom ESLint rule for each top-level directory in `.eslintrc.json`:

https://github.com/WordPress/gutenberg/blob/e694f6a99b7ce01bb3586798e1af6c662a7a4895/.eslintrc.json#L95-L125

aduth reviewed (commented):
I agree with general direction that this helps us toward consolidating `editor` and `blocks` directory as also noted in #2761. There's a few open questions remaining on state responsible for editor vs. `edit-post` (e.g. `currentPost` remains in editor state) that we could consider refactoring further in subsequent pull requests.

The addition of `storeKey` to each connected component is quite verbose. I think we should consider either a `connect` custom to each top-level module which serves as a wrapper to React-Redux's `connect` passing the `storeKey` automatically. Alternatively, we could seek to enhance our own `query` HoC to support all of the same functionality as `connect` and then it would be not necessary.

Generally difficult to review some of the more granular changes, as while some files are moved verbatim, others include revisions.

youknowriad committed:
Changes per review

youknowriad committed:
update renderBlockMenu API

youknowriad commented:
> Alternatively, we could seek to enhance our own query HoC to support all of the same functionality as connect and then it would be not necessary.

If we're fine exposing all selectors to third-party modules, we can get rid of the `connect` usage.

--

I addressed most of the feedback. Thanks @aduth

aduth commented on editor/components/block-list/block.js:
```diff
@@ -433,7 +433,7 @@ export class BlockListBlock extends Component {
 				<BlockDropZone index={ order } />
 				{ ( showUI || isHovered ) && <BlockMover uids={ [ block.uid ] } /> }
-				{ ( showUI || isHovered ) && <BlockSettingsMenu uids={ [ block.uid ] } /> }
+				{ ( showUI || isHovered ) && <BlockSettingsMenu uids={ [ block.uid ] } renderBlockMenu={ renderBlockMenu } /> }
```
This change with `renderBlockMenu` broke the multi-select controls dropdown:

1. Create a multi-selection range
2. Open "More Options" menu

![multi-select missing](https://user-images.githubusercontent.com/1779930/35481390-98fec504-041a-11e8-97d1-9ddfe0f8c6df.png)

youknowriad replied:
Good catch, we should cherry-pick the fix from the templates branch to its own PR

aduth commented:
Nitpick: To the naming of this module, it stands out to me as inconsistent with our general naming, not only because it's intended as a superset of functionality for the base `editor` (but naming does not imply this), but also that other modules generally tend to be nouns (`editor`, `blocks`, `components`), whereas `edit-post` is more an action/verb.

My suggested alternative would be `post-editor`. What do you think?

aduth commented:
Also, we should plan to fix the assignment of the global to avoid `wp[ 'edit-post' ]` and instead as `wp.editPost`. I agree it wasn't a blocker for this pull request, but I don't think we should settle on the non-camelCased global property.

youknowriad commented:
cc @mtias for the name of the module.

About the global, this might mean we need to find another way to assign these variables (instead of relying on the library config of webpack)

aduth commented on editor/components/block-list/index.js:
```diff
@@ -245,10 +246,10 @@ class BlockList extends Component {
 
 		return (
-			<div>
+			<BlockSelectionClearer>
```
Why was this changed from a `div` to `BlockSelectionClearer` ?

youknowriad replied:
Because this was handled in the parent component by using a ref to this div there. I just made this component generic and used it in both the `VisualEditor` component and the inner component `BlockList`

jasmussen commented:
I think this PR regressed the color scheme styling of the tabs in the sidebar:

<img width="218" alt="screen shot 2018-01-31 at 10 20 42" src="https://user-images.githubusercontent.com/1204802/35614810-936e03a4-0670-11e8-9888-d9ea7bdbf8d1.png">

As you can see from the Publish button, it's supposed to be red, like other tab buttons:

<img width="402" alt="screen shot 2018-01-31 at 10 22 17" src="https://user-images.githubusercontent.com/1204802/35614823-a25b9c5a-0670-11e8-9391-7cc2ed875d90.png">

It should be a fairly easy fix, renaming some CSS classes in edit-post/assets/stylesheets/_admin-schemes.scss.

Do you want to do it, @youknowriad, or would you like me to do it? I have time if you're busy.
